### PR TITLE
Pluggable optimizer framework — Phase 3A (agent-running infra)

### PR DESCRIPTION
**Pull request by egonSchiele**
## Summary

Phase 3A of the pluggable optimizer framework: the **agent-running infrastructure**, added as
additive, fully-tested services. **Nothing is wired into the live optimizer yet — zero behavior
change** (greedy still uses the Phase 1 pairwise path). The greedy → pointwise migration is
Phase 3B.

- **Refactor `runAgencyNode`** (`lib/cli/util.ts`) — extracts the general "run an `.agency` node" core (compile/resolve → render evaluate script → spawn `node` → parse results) from `executeNodeAsync`. The test-LLM coupling (`AGENCY_USE_TEST_LLM_PROVIDER`/`useTestLLMProvider` → `AGENCY_LLM_MOCKS`) now lives only in `executeNodeAsync`, which delegates to the core. Existing callers are unchanged.
- **`AgencyRunner`** (class) — `run()` executes the agent under test; `runStructured()` runs a judge/proposer node and zod-validates its return. A `runNode` seam makes it unit-testable.
- **`GraderInput += runAgency`** — graders can now invoke a judge `.agency` file.
- **`LlmJudge`** grader — scalar or binary verdicts from a judge `.agency` file.
- **`EvalCache`** — memoizes one `AgentRun` per `(workspace, input)` (the "run once, grade many" guarantee).
- **`WorkspaceManager`** — per-iteration workspace dirs (fork/read/write/applyFiles).

## Test Plan

- [x] `pnpm test:run lib/optimize lib/cli/util.test.ts lib/cli/eval` — 194 passed across 24 files.
- [x] `pnpm typecheck` — clean.
- [x] `pnpm run lint:structure` — clean.
- [x] Zero behavior change: the Phase 1 pairwise-greedy CLI/loop tests are unchanged and green; the refactor's consumers (mutator, eval runner) still pass through `executeNodeAsync` → `runAgencyNode`.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: packages/agency-lang/lib/optimize/grading/llmJudge.test.ts [hunk 1]
```diff
@@ -0,0 +1,26 @@
+import { describe, expect, it } from "vitest";
+
+import { AgencyRunner } from "./agencyRunner.js";
+import { LlmJudge } from "./llmJudge.js";
+import type { GraderInput, Input } from "./types.js";
+
+const gi = (verdict: { score?: number; pass?: boolean; reasoning: string }): GraderInput => {
```
Comment (egonSchiele) [inline packages/agency-lang/lib/optimize/grading/llmJudge.test.ts]: `gi` is a terrible name

File: packages/agency-lang/lib/optimize/grading/llmJudge.ts [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+
+/** Grades an output by running a judge .agency file and reading its structured verdict. */
+export class LlmJudge extends BaseGrader {
+  protected readonly defaultName = "llm-judge";
+  constructor(protected readonly options: LlmJudgeOptions) {
+    super(options);
+  }
+
+  protected async _run({ input, run, runAgency }: GraderInput): Promise<Grade> {
+    const goal = String(getPath(input, this.options.goalPath ?? ["metadata", "goal"]) ?? "");
```
Comment (egonSchiele) [inline packages/agency-lang/lib/optimize/grading/llmJudge.ts]: I'm curious how effective you think an LLM judge would be with no goal. What do you expect the LLM judge to do in that scenario?

File: packages/agency-lang/lib/optimize/evalCache.ts [hunk 3]
```diff
@@ -0,0 +1,14 @@
+import type { AgentRun } from "./grading/types.js";
+
+/** Memoizes one AgentRun per (workspaceKey, inputId). Null-prototype: keys derive from user inputs. */
+export class EvalCache {
+  private readonly runs: Record<string, Promise<AgentRun>> = Object.create(null);
+
+  get(workspaceKey: string, inputId: string, produce: () => Promise<AgentRun>): Promise<AgentRun> {
+    const key = `${workspaceKey} ${inputId}`;
```
Comment (egonSchiele) [inline packages/agency-lang/lib/optimize/evalCache.ts]: replace the space with a hyphen

File: packages/agency-lang/lib/optimize/grading/llmJudge.ts (hunk 2 above)
Comment (egonSchiele) [reply packages/agency-lang/lib/optimize/grading/llmJudge.ts]: You're right — it would do nothing useful. With an empty goal the judge agent is asked to grade the output against no criterion, so its verdict is meaningless (and worse, it looks like a real score). I treated a missing goal as silent-success-ish, which is exactly the wrong default.

Fixed in bf0d4d65: `_run` now throws a clear error when the goal doesn't resolve on the input (undefined/null/empty-after-trim) — `"<name>: no goal found at <path> on input <id>; an LLM judge needs a goal."` — with a test. So a misconfigured judge fails loudly at the input rather than emitting a bogus score.

File: packages/agency-lang/lib/optimize/grading/llmJudge.test.ts (hunk 1 above)
Comment (egonSchiele) [reply packages/agency-lang/lib/optimize/grading/llmJudge.test.ts]: Renamed the gi helper to graderInput across the grading tests in bf0d4d65.

File: packages/agency-lang/lib/optimize/workspace.ts [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+
+  read(ws: Workspace, relPath: string): string {
+    return fs.readFileSync(path.join(ws.dir, relPath), "utf8");
+  }
+
+  write(ws: Workspace, relPath: string, content: string): void {
+    const abs = path.join(ws.dir, relPath);
+    fs.mkdirSync(path.dirname(abs), { recursive: true });
+    fs.writeFileSync(abs, content);
+  }
```
[the remark this answers is not in the document]
Comment (egonSchiele) [reply packages/agency-lang/lib/optimize/workspace.ts:28]: Fixed in bf0d4d65: read/write now resolve via a private resolveWithin() that throws if the resolved path escapes the workspace root (covers .. and absolute paths). Added a traversal test.

File: packages/agency-lang/lib/optimize/evalCache.ts [hunk 5]
```diff
@@ -0,0 +1,14 @@
...
+    if (!Object.hasOwn(this.runs, key)) {
+      this.runs[key] = produce();
+    }
+    return this.runs[key];
+  }
```
[the remark this answers is not in the document]
Comment (egonSchiele) [reply packages/agency-lang/lib/optimize/evalCache.ts:18]: Fixed in bf0d4d65: EvalCache now uses a nested store keyed by workspace then input — no separator at all, so the two components cannot collide. Added a regression test for the boundary case.

File: packages/agency-lang/lib/optimize/grading/agencyRunner.ts [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+  }
+
+  private async exec(agencyFile: string, nodeName: string, args: Record<string, JSON>): Promise<{ data: unknown }> {
+    const scratchDir = fs.mkdtempSync(path.join(os.tmpdir(), "agency-runner-"));
+    try {
+      const config = { ...this.config };
+      delete config.distDir;
+      const argsString = Object.values(args).map((v) => globalThis.JSON.stringify(v)).join(", ");
+      return await this.runNode({ config, agencyFile, nodeName, hasArgs: argsString.length > 0, argsString, scratchDir, quietCompile: true });
+    } finally {
```
[the remark this answers is not in the document]
Comment (egonSchiele) [reply packages/agency-lang/lib/optimize/grading/agencyRunner.ts]: Fixed in bf0d4d65: run() and runStructured() now take positional args (JSON[]) instead of Object.values(Record), so call order is explicit. (The named->positional mapping for the agent-under-test is handled in BaseOptimizer.runAgent in 3B, where the node param order is known.)

File: packages/agency-lang/lib/optimize/evalCache.ts (hunk 3 above)
Comment (egonSchiele) [reply packages/agency-lang/lib/optimize/evalCache.ts]: Went with a nested map rather than a hyphen: a hyphen still collides because workspace keys are themselves ws-N (already contain hyphens), e.g. (ws-1, 2-x) vs (ws-1-2, x) both flatten to ws-1-2-x. Nesting by workspace then input removes the separator entirely. (bf0d4d65)
